test/cpp/tensorexpr/test_loopnest.cpp

```diff
@@ -546,5 +547,224 @@ void testScheduleDynamicShape2D() {
+  ASSERT_EQ(bounds_info_map.size(), 2);
+  ASSERT_EQ(bounds_info_map.at(a.data()).kind, kLoad);
+  verifyConstBounds(bounds_info_map.at(a.data()), {{0, 109}});
+  ASSERT_EQ(bounds_info_map.at(b->buf()).kind, kStore);
+  verifyConstBounds(bounds_info_map.at(b->buf()), {{0, 99}});
+}
+
+void testBoundsInference_4() {
+  // Verify that bounds inference works for the following example:
+  //
+  // for y in 0..200:
+  //   for x in 0..320:
+  //     b[y,x] = x*y
+  // for y in 0..200:
+  //   for x in 0..320:
+  //     c[y,x] = a[y,x] * b[y,x]
+  KernelScope kernel_scope;
+  ExprHandle W(320);
+  ExprHandle H(200);
+  Buffer a(BufHandle("a", {H, W}), kFloat);
+  Tensor* b = Compute(
+      "b", {{H, "y"}, {W, "x"}}, [&](const VarHandle& y, const VarHandle& x) {
+        return x * y;
+      });
+  Tensor* c = Compute(
+      "c", {{H, "y"}, {W, "x"}}, [&](const VarHandle& y, const VarHandle& x) {
+        return a(y, x) * b->call(y, x);
+      });
+  LoopNest l({c});
+  std::vector<For*> loops = l.getLoopStmtsFor(c);
+  Stmt* body = l.getLoopBodyFor(c);
+  {
+    // Infer bounds on the top-level loop scope
+    const std::vector<TensorAccessBoundsInfo>& bounds_info =
+        inferBounds(loops[0]);
+    auto bounds_info_map = convertBoundsInfoToMap(bounds_info);
+
+    ASSERT_EQ(bounds_info_map.at(a.data()).kind, kLoad);
+    verifyConstBounds(bounds_info_map.at(a.data()), {{0, 199}, {0, 319}});
+
+    ASSERT_EQ(bounds_info_map.at(b->buf()).kind, kLoad);
+    verifyConstBounds(bounds_info_map.at(b->buf()), {{0, 199}, {0, 319}});
+
+    ASSERT_EQ(bounds_info_map.at(c->buf()).kind, kStore);
+    verifyConstBounds(bounds_info_map.at(c->buf()), {{0, 199}, {0, 319}});
+  }
+  {
+    // Infer bounds on the inner loop scope
+    const std::vector<TensorAccessBoundsInfo>& bounds_info =
+        inferBounds(loops[1]);
+    auto bounds_info_map = convertBoundsInfoToMap(bounds_info);
+
+    ASSERT_EQ(bounds_info_map.at(a.data()).kind, kLoad);
+    verifyConstBounds(bounds_info_map.at(a.data()), {{-1, -1}, {0, 319}});
+
+    ASSERT_EQ(bounds_info_map.at(b->buf()).kind, kLoad);
+    verifyConstBounds(bounds_info_map.at(b->buf()), {{-1, -1}, {0, 319}});
+
+    ASSERT_EQ(bounds_info_map.at(c->buf()).kind, kStore);
+    verifyConstBounds(bounds_info_map.at(c->buf()), {{-1, -1}, {0, 319}});
+  }
+  {
+    // Infer bounds on the inner loop body's scope
+    const std::vector<TensorAccessBoundsInfo>& bounds_info =
+        inferBounds(body);
+    auto bounds_info_map = convertBoundsInfoToMap(bounds_info);
+
+    ASSERT_EQ(bounds_info_map.at(a.data()).kind, kLoad);
+    verifyConstBounds(bounds_info_map.at(a.data()), {{-1, -1}, {-1, -1}});
+
+    ASSERT_EQ(bounds_info_map.at(b->buf()).kind, kLoad);
+    verifyConstBounds(bounds_info_map.at(b->buf()), {{-1, -1}, {-1, -1}});
+
+    ASSERT_EQ(bounds_info_map.at(c->buf()).kind, kStore);
+    verifyConstBounds(bounds_info_map.at(c->buf()), {{-1, -1}, {-1, -1}});
+  }
+}
+
+void testBoundsInference_5() {
```

Comment:
Another important test case is where the same source tensor is used multiple times in the target tensor, potentially through different path: 

a = f(x, y)
b = a(x, y)
c = b(x+1, y)
d = a(x, y+1) + c(x, y)
compute(d)
a.compute_at(d, x)

Reply:
That comment is probably to the next PR, not this one?

Reply:
I added this test to `computeAt` tests.